### PR TITLE
Update payment method suggestion endpoint

### PR DESCRIPTION
**Pull request by louwie17**
This updates the payment method suggestions logic to use the new `payment-method/suggestions.json` endpoint that follows the same data format as our other specs. This allows the country and plugin validation to be moved to the specs, instead of being handled on the client.

I did have to update the endpoint a little bit, there is a separate PR for those changes.

### Detailed test instructions:

- Load this branch and also the `update/payment-recommendations-endpoint` branch in woocommerce.com. 
- Make sure you have market place suggestions enabled.
- Go to **WooCommerce > Settings > Payments**
- Check if the **Recommended ways to get paid** shows up below the payment table with WcPay, Stripe, and Paypal (unless you have one installed).
- Click on **Get started** for one and make sure it installs correctly and redirects to the set up screen.
- Go back to the **Settings > Payments** screen, the card should now show up without the installed & activated plugin.
- Set your WooCommerce store to a WC Pay unsupported country (with WC Pay disabled ) you could use Brazil.
- Go back to **Settings > Payments**
- Make sure WC Pay doesn't show up in the list
- Set your WooCommerce store back to a supported WC Pay country
- Change the return value [here](https://github.com/woocommerce/woocommerce-admin/blob/main/src/Features/WcPayPromotion/Init.php#L122) to `true`
- Reload the **Settings > Payments** screen, make sure that WooCommerce Payments shows in the payment method list with an **Install** button and the card does not display WC Pay.
- Open your console and make sure the `wcadmin_settings_payments_recommendations_pageview` track fired with props for each gateway`_displayed`, if you don't see the tracks run `localStorage.setItem( 'debug', 'wc-admin:*' );` in your console and refresh the page.
- Click on the 3 dots of the card and select **Hide this**
- The card should now be hidden, refresh the page and make sure this is persisted
- Look in your console again, the `wcadmin_settings_payments_recommendations_pageview` track should still be fired, but just with the woocommerce_payments prop for the experiment.

### CODE REVIEW

File: src/PaymentMethodSuggestionsDataSourcePoller.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
+<?php
```
Comment (louwie17) [inline src/PaymentMethodSuggestionsDataSourcePoller.php:1]: This replaces the `PaymentPlugins.php` file.

File: client/payments/payment-recommendations.tsx [hunk 2]
```diff
@@ -24,64 +22,23 @@ import { getAdminLink } from '@woocommerce/wc-admin-settings';
...
-	const countryCode =
-		settings && settings.woocommerce_default_country
-			? getCountryCode( settings.woocommerce_default_country )
-			: null;
-	const countrySupported = countryCode
-		? isWCPaySupported( countryCode )
-		: false;
-	const isRequestingOptions = isResolvingOption( 'getOption', [
-		DISMISS_OPTION,
-	] );
```
Comment (louwie17) [inline client/payments/payment-recommendations.tsx:68]: We do not need the dismiss option anymore and don't need the country info either. All of this is handled on the back-end now, so we only do one request.

File: packages/data/src/plugins/actions.ts [hunk 3]
```diff
@@ -348,6 +349,36 @@ export function setRecommendedPlugins(
...
+export function* dismissRecommendedPlugins( type: RecommendedTypes ) {
+	if ( ! SUPPORTED_TYPES.includes( type ) ) {
+		return [];
+	}
+	const plugins: Plugin[] = yield resolveSelect(
+		STORE_NAME,
+		'getRecommendedPlugins',
+		type
+	);
+	yield setRecommendedPlugins( type, [] );
```
Comment (louwie17) [inline packages/data/src/plugins/actions.ts:362]: Optimistically dismiss the recommend plugins, if the below request fails we set the recommended plugins back

File: src/DataSourcePoller.php [hunk 4]
```diff
@@ -105,6 +110,7 @@ public function get_specs_from_data_sources() {
 			$this->read_specs_from_data_sources();
 			$specs = get_transient( $this->args['transient_name'] );
 		}
+		$specs = apply_filters( self::FILTER_NAME_SPECS, $specs, $this->id );
```
Comment (louwie17) [inline src/DataSourcePoller.php:113]: Allow specs to be filtered before returning them.
I am using this inside the `WcPayPromotion/Init.php` file.

**Comment by joelclimbsthings** [comment]
Hey @louwie17 ! 

Before I dive to deep into this, just wanted to ensure I was testing correctly. 

I'm assuming that after checking out the branch you're referencing from woocommerce.com I need to change the URL on [this line](https://github.com/woocommerce/woocommerce-admin/blob/aed8aa65bc7190c129b4d6e78978d5a23fd235bd/src/PaymentMethodSuggestionsDataSourcePoller.php#L24) to reference my local?

And also, delete the transient `_transient_woocommerce_admin_payment_method_suggestions_specs`?

Having done that. I've got marketplace suggestions on and my country set to USA (Oregon), and I don't see WCPay on the Payment settings. I don't have it installed at all:

image

Just wanted to make sure I don't have anything incorrect with my setup.

**Comment by mattsherman** [comment]
> I'm assuming that after checking out the branch you're referencing from woocommerce.com I need to change the URL on [this line](https://github.com/woocommerce/woocommerce-admin/blob/aed8aa65bc7190c129b4d6e78978d5a23fd235bd/src/PaymentMethodSuggestionsDataSourcePoller.php#L24) to reference my local?

Favor adding a filter in a test mu-plugin such the following, over changing source code directly when reviewing/testing things (un-tested, but copied from a similar filter I have used for other similar testing):

```
add_filter(
	'data_source_poller_data_sources',
	function( $data_sources, $data_source_id ) {

		if ( $data_source_id === 'payment_method_suggestions' ) {
			// Use the default IP address of the Vagrant host; there needs to be a better way to handle this!
			return array( 'http://10.0.2.2/wp-json/wccom/payment-gateway-suggestions/1.0/payment-method/suggestions.json );
		}

		return $data_sources;
	}, 10, 2
);

```

**Comment by louwie17** [comment]
Thanks for testing this @joelclimbsthings 

> Having done that. I've got marketplace suggestions on and my country set to USA (Oregon), and I don't see WCPay on the Payment settings. I don't have it installed at all:

It could be that your are hitting the `treatment` experiment of promoting WC Pay, if you are not seeing it in the card do you see it in the Payment methods table above the card instead with an `Install` button?

File: packages/data/src/plugins/actions.ts [hunk 5]
```diff
@@ -348,6 +349,36 @@ export function setRecommendedPlugins(
 	};
 }
 
+const SUPPORTED_TYPES = [ 'payments' ];
+export function* dismissRecommendedPlugins( type: RecommendedTypes ) {
```
Comment (joelclimbsthings) [inline packages/data/src/plugins/actions.ts:353]: Is there a good way to handle the typescript return type with these generator functions? My linting is going into 🎄 mode on this file.

**Review by joelclimbsthings — APPROVED** [review]
Thanks @louwie17 ! Love these changes. Once I resolved the experiment and such, everything worked as expected for the most part. 

For the tracks event `wcadmin_settings_payments_recommendations_pageview`, should it include wcPay as `displayed: true` when it's only shown in the payment method list, and not the card below? This was the case on my end, so I wanted to ensure that that's expected.

Other than that :shipit:

File: packages/data/src/plugins/actions.ts (hunk 5 above)
Comment (louwie17) [reply packages/data/src/plugins/actions.ts:353]: I am actually not a 100% sure, as I don't think we set the return types explicitly anywhere. Also from what I gather that whenever you call a function with `yield someFunction()` it actually returns an `unknown` type, so you did have to set the type on the returned value. 
Usually I would go see what Gutenberg does for this, but their's is still vanilla JS.

**Pull request merged**